### PR TITLE
Catch unhandled exception and avoid crash on test host exit

### LINKED ISSUE #4222: Test host crashes at cleanup with COMException
On our build agents, we occasionally observe red test execution with all test cases green. It seems to be that the crash happens at test cleanup:

```
Unhandled exception. System.Runtime.InteropServices.COMException (0x80070006): The handle is invalid. (0x80070006 (E_HANDLE))
at System.Runtime.InteropServices.Marshal.ThrowExceptionForHR(Int32 errorCode)
at Interop.Kernel32.ProcessWaitHandle..ctor(SafeProcessHandle processHandle)
at System.Diagnostics.Process.UpdateHasExited()
at System.Diagnostics.Process.get_HasExited()
at Microsoft.VisualStudio.TestPlatform.PlatformAbstractions.ProcessHelper.TryGetExitCode(Object process, Int32& exitCode)
at Microsoft.TestPlatform.TestHostProvider.Hosting.TestHostManagerCallbacks.ExitCallBack(IProcessHelper processHelper, Object process, StringBuilder testHostProcessStdError, Action`1 onHostExited)
at Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Hosting.DotnetTestHostManager.<get_ExitCallBack>b__37_0(Object process)
at Microsoft.VisualStudio.TestPlatform.PlatformAbstractions.ProcessHelper.<>c__DisplayClass1_0.<<LaunchProcess>b__3>d.MoveNext()
--- End of stack trace from previous location ---
at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__128_1(Object state)
at System.Threading.QueueUserWorkItemCallbackDefaultContext.Execute()
at System.Threading.ThreadPoolWorkQueue.Dispatch()
at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart()
at System.Threading.Thread.StartCallback()
```

The crucial part here is that `Process.get_HasExited` [is documented](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.process.hasexited?view=net-7.0#system-diagnostics-process-hasexited) to crash with InvalidOperationException, Win32Exception or NotSupportedException, none of which includes the COMException.

I wonder whether there is a missing call to `Process.WaitForExit()` as documented in the same location for asynchronous event handlers, but in general I have no clue what is going on. I tend to think that there is a race condition in the `Process` class:

image

This issue occurred 3 times in the last 600 analyzed runs.

### PR DESCRIPTION
**Pull request by MarcoRossignoli**
fixes https://github.com/microsoft/vstest/issues/4222

### CODE REVIEW

File: src/Microsoft.TestPlatform.TestHostProvider/Hosting/TestHostManagerCallbacks.cs [hunk 1]
```diff
@@ -14,6 +15,8 @@ namespace Microsoft.TestPlatform.TestHostProvider.Hosting;
 
 internal class TestHostManagerCallbacks
 {
+    const int E_HANDLE = unchecked((int)0x80070006);
```
Comment (Evangelink) [inline src/Microsoft.TestPlatform.TestHostProvider/Hosting/TestHostManagerCallbacks.cs:18]: Nit: explicit modifier?

```suggestion
    private const int E_HANDLE = unchecked((int)0x80070006);
```

File: test/Microsoft.TestPlatform.TestHostProvider.UnitTests/Hosting/TestHostManagerCallbacksTests.cs [hunk 2]
```diff
@@ -94,4 +98,103 @@ public void ErrorReceivedCallbackShouldAppendNewLineApproprioritlyWhenReachingMa
...
+    public void ErrorReceivedCallbackShouldNotCrashIfInvalidProcessHandle()
+    {
+        var handleProcessHelper = new InvalidHandleProcessHelper();
+        bool onHostExitedCalled = false;
+        TestHostManagerCallbacks.ExitCallBack(handleProcessHelper, null, new StringBuilder(),
+        hostProviderEventArgs =>
+        {
+            onHostExitedCalled = true;
+            Assert.AreEqual(-1, hostProviderEventArgs.ErrroCode);
+        });
```
Comment (Evangelink) [inline test/Microsoft.TestPlatform.TestHostProvider.UnitTests/Hosting/TestHostManagerCallbacksTests.cs:112]: Nit: fix indentation.

```suggestion
        TestHostManagerCallbacks.ExitCallBack(handleProcessHelper, null, new StringBuilder(),
            hostProviderEventArgs =>
            {
                onHostExitedCalled = true;
                Assert.AreEqual(-1, hostProviderEventArgs.ErrroCode);
            });
```

File: test/Microsoft.TestPlatform.TestHostProvider.UnitTests/Hosting/TestHostManagerCallbacksTests.cs [hunk 3]
```diff
@@ -94,4 +98,103 @@ public void ErrorReceivedCallbackShouldAppendNewLineApproprioritlyWhenReachingMa
...
+
+        Assert.IsTrue(handleProcessHelper.TryGetExitCodeCalled, "TryGetExitCodeCalled was not called");
+        Assert.IsTrue(onHostExitedCalled, "onHostExited was not called");
+    }
+
+    class InvalidHandleProcessHelper : IProcessHelper
```
Comment (Evangelink) [inline test/Microsoft.TestPlatform.TestHostProvider.UnitTests/Hosting/TestHostManagerCallbacksTests.cs:118]: Nit: Explicit modifier

```suggestion
    private class InvalidHandleProcessHelper : IProcessHelper
```

File: test/Microsoft.TestPlatform.TestHostProvider.UnitTests/Hosting/TestHostManagerCallbacksTests.cs [hunk 4]
```diff
@@ -94,4 +98,103 @@ public void ErrorReceivedCallbackShouldAppendNewLineApproprioritlyWhenReachingMa
 
         Assert.AreEqual("1234" + Environment.NewLine.Substring(0, 1), _testHostProcessStdError.ToString());
     }
+
+    [TestMethod]
+    public void ErrorReceivedCallbackShouldNotCrashIfInvalidProcessHandle()
+    {
+        var handleProcessHelper = new InvalidHandleProcessHelper();
```
Comment (Evangelink) [inline test/Microsoft.TestPlatform.TestHostProvider.UnitTests/Hosting/TestHostManagerCallbacksTests.cs:105]: I believe you could simply use a mock here.

Comment (MarcoRossignoli) [reply test/Microsoft.TestPlatform.TestHostProvider.UnitTests/Hosting/TestHostManagerCallbacksTests.cs:105]: yeah....I start to hate mocks...fakes are so simple  will do.

**Review by Evangelink — APPROVED** [review]

**Pull request merged**